modules/rhsm_manifest.py

```diff
@@ -0,0 +1,264 @@
+    resp, info = fetch_portal(module, path, 'GET')
+    manifests = json.loads(resp.read())
+    if module.params['name']:
+        attr = 'name'
+    if module.params['uuid']:
+        attr = 'uuid'
+    manifest = filter(lambda m: m[attr] == module.params[attr], manifests)
+    if manifest:
+        return manifest[0], False
+    else:
+        return create_manifest(module), True
+
+
+def get_owner(module):
+    path = "/subscription/users/%s/owners" % (module.params['username'])
+    resp, info = fetch_portal(module, path, 'GET')
+    return json.loads(resp.read())[0]['key']
+
+
+def get_subs(module, manifest):
+    path = "/subscription/consumers/%s/entitlements" % (manifest['uuid'])
+    resp, info = fetch_portal(module, path, 'GET')
+    all_subs = json.loads(resp.read())
+    subs = filter(lambda s: s['pool']['id'] == module.params['pool_id'], all_subs)
+    return subs
+
+
+def get_remove_or_attach_sub(module, manifest):
+    changed = False
+    subs = get_subs(module, manifest)
+    if subs:
+        if module.params['state'] == 'present':
+            sub_quantity = sum(map(lambda s: s['quantity'], subs))
+            while sub_quantity > module.params['quantity']:
+                difference = sub_quantity - module.params['quantity']
+                remove_sub(module, manifest, subs[0])
+                changed = True
+                subs = get_subs(module, manifest)
+                sub_quantity = sum(map(lambda s: s['quantity'], subs))
+            if sub_quantity < module.params['quantity']:
+                difference = module.params['quantity'] - sub_quantity
+                attach_sub(module, manifest, difference)
+                changed = True
+        elif module.params['state'] == 'absent':
+            for sub in subs:
+                remove_sub(module, manifest, sub)
+            changed = True
+    elif module.params['state'] == 'present':
+        attach_sub(module, manifest, module.params['quantity'])
+        changed = True
+    return changed
+
+
+def remove_sub(module, manifest, sub):
+    path = "/subscription/consumers/%s/entitlements/%s" % (manifest['uuid'], sub['id'])
+    resp, info = fetch_portal(module, path, 'DELETE')
+
+
+def attach_sub(module, manifest, quantity):
+    path = "/subscription/consumers/%s/entitlements?pool=%s&quantity=%s" % (manifest['uuid'], module.params['pool_id'], quantity)
+    resp, info = fetch_portal(module, path, 'POST')
+
+
+def export_manifest(module, manifest):
+    path = "/subscription/consumers/%s/export" % (manifest['uuid'])
+    try:
+        resp, info = fetch_portal(module, path, 'GET', accept_header='application/zip')
+        f = open(module.params['path'], 'w')
+        # Read 1kb at a time to save on ram
+        while True:
+            data = resp.read(65536)  # 64K
+            if data == "":
+                break  # End of file, break while loop
+            f.write(data)
+        f.close()
+    except Exception:
+        e = get_exception()
```

Comment:
Any reason you're not using ```except Exception as e```? Is this an ansible thing?

Reply:
👍 `get_exception()` is an ansible thing, https://github.com/theforeman/foreman-ansible-modules/pull/71/files#diff-88ef7a0ec70e64151c3a125a04cb3b36R114